pkg/util/scheduler/queue_test.go

```diff
@@ -0,0 +1,635 @@
+
+		// Dequeue items
+		for i := 0; i < numItems; i++ {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				dequeueCtx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+				defer cancel()
+				runnable, ok, err := q.Dequeue(dequeueCtx)
+				require.NoError(t, err, "Dequeue should succeed")
+				require.True(t, ok, "Dequeue should return ok=true")
+				require.NotNil(t, runnable, "Dequeued runnable should not be nil")
+			}()
+		}
+
+		wg.Wait()
+
+		// Let's simplify the dequeue check: Dequeue sequentially after enqueueing.
+		qSimple := NewQueue(QueueOptionsWithDefaults(nil))
+		require.NoError(t, qSimple.StartAsync(context.Background()), "Queue should start")
+		require.NoError(t, qSimple.AwaitRunning(context.Background()), "Queue should be running")
+
+		for i := 0; i < numItems; i++ {
+			err := qSimple.Enqueue(ctx, tenantID, func() {})
+			require.NoError(t, err)
+		}
+		require.Equal(t, numItems, qSimple.Len(), "Queue length after enqueue (simple)")
+		require.Equal(t, 1, qSimple.ActiveTenantsLen(), "Active tenants after enqueue (simple)")
+
+		for i := 0; i < numItems; i++ {
+			dequeueCtx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+			runnable, ok, err := qSimple.Dequeue(dequeueCtx)
+			cancel() // Cancel context after use
+			require.NoError(t, err, "Dequeue %d should succeed (simple)", i)
+			require.True(t, ok, "Dequeue %d should return ok=true (simple)", i)
+			require.NotNil(t, runnable, "Dequeued runnable %d should not be nil (simple)", i)
+		}
+
+		require.Equal(t, 0, qSimple.Len(), "Queue length after dequeue (simple)")
+		require.Equal(t, 0, qSimple.ActiveTenantsLen(), "Active tenants after dequeue (simple)")
+
+		// Check dequeue on empty queue times out
+		dequeueCtx, cancel := context.WithTimeout(ctx, 50*time.Millisecond)
+		_, ok, err := qSimple.Dequeue(dequeueCtx)
+		cancel()
+		require.ErrorIs(t, err, context.DeadlineExceeded, "Dequeue on empty queue should time out")
+		require.False(t, ok, "Dequeue on empty queue should return ok=false")
+
+		// Stop the queue
+		qSimple.StopAsync()
+		require.NoError(t, qSimple.AwaitTerminated(context.Background()), "Queue should stop")
+	})
+
+	t.Run("RoundRobinBetweenTenants", func(t *testing.T) {
+		t.Parallel()
+
+		ctx := context.Background()
+		q := NewQueue(QueueOptionsWithDefaults(nil))
+		require.NoError(t, q.StartAsync(ctx), "Queue should start")
+		require.NoError(t, q.AwaitRunning(ctx), "Queue should be running")
+
+		tenantA := "tenant-a"
+		tenantB := "tenant-b"
+
+		// We'll use a very small test with just 2 items per tenant
+		// to reduce the chance of timeouts or other issues
+
+		// Enqueue items in tenant order: A, B, A, B
+		// Each item will record its tenant ID when executed
+		var results []string
+		var resultsMu sync.Mutex
+
+		makeRunnable := func(id string) func() {
+			return func() {
+				resultsMu.Lock()
+				results = append(results, id)
+				resultsMu.Unlock()
+			}
+		}
+
+		// First tenant A, then B, then A, then B
```

Comment:
Would it make more sense to enqueue tenant A first, then tenant B? We want to see fair tenant usage during dequeue, but now enqueue and dequeue orders are the same, so it doesn't show any reordering due to fairness.